design/src/rfcs/rfc0016_body_callback_apis_v2.md

```diff
@@ -0,0 +1,389 @@
+            .expect("base64 will always produce valid header values from checksums");
+
+        header_map.insert(key, value);
+
+        Ok(Some(header_map))
+    }
+
+    // Size of the checksum in bytes
+    fn size() -> u64 {
+        20
+    }
+}
+
+#[derive(Debug, Default)]
+struct Sha256Callback {
+    hasher: sha2::Sha256,
+}
+
+impl Sha256Callback {
+    fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
+        self.hasher.write_all(bytes)?;
+
+        Ok(())
+    }
+
+    fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
+        let mut header_map = HeaderMap::new();
+        let key = HeaderName::from_static(SHA_256_HEADER_NAME);
+        // We clone the hasher because `Hasher::finalize` consumes `self`
+        let hash = self.hasher.clone().finalize();
+        let value = HeaderValue::from_str(&base64::encode(&hash[..]))
+            .expect("base64 will always produce valid header values from checksums");
+
+        header_map.insert(key, value);
+
+        Ok(Some(header_map))
+    }
+
+    // Size of the checksum in bytes
+    fn size() -> u64 {
+        32
+    }
+}
+
+enum Inner {
+    Crc32(Crc32Callback),
+    Crc32c(Crc32cCallback),
+    Sha1(Sha1Callback),
+    Sha256(Sha256Callback),
+}
+
+pub struct ChecksumCallback(Inner);
+
+impl ChecksumCallback {
+    pub fn new(checksum_algorithm: &str) -> Self {
+        if checksum_algorithm.eq_ignore_ascii_case(CRC_32_NAME) {
+            Self(Inner::Crc32(Crc32Callback::default()))
+        } else if checksum_algorithm.eq_ignore_ascii_case(CRC_32_C_NAME) {
+            Self(Inner::Crc32c(Crc32cCallback::default()))
+        } else if checksum_algorithm.eq_ignore_ascii_case(SHA_1_NAME) {
+            Self(Inner::Sha1(Sha1Callback::default()))
+        } else if checksum_algorithm.eq_ignore_ascii_case(SHA_256_NAME) {
+            Self(Inner::Sha256(Sha256Callback::default()))
+        } else {
+            panic!("unsupported checksum algorithm '{}'", checksum_algorithm)
+        }
+    }
+
+    pub fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
+        match &mut self.0 {
+            Inner::Crc32(ref mut callback) => callback.update(bytes)?,
+            Inner::Crc32c(ref mut callback) => callback.update(bytes)?,
+            Inner::Sha1(ref mut callback) => callback.update(bytes)?,
+            Inner::Sha256(ref mut callback) => callback.update(bytes)?,
+        };
+
+        Ok(())
+    }
+
+    pub fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
+        match &self.0 {
+            Inner::Sha256(callback) => callback.trailers(),
+            Inner::Crc32c(callback) => callback.trailers(),
+            Inner::Crc32(callback) => callback.trailers(),
+            Inner::Sha1(callback) => callback.trailers(),
+        }
+    }
+
+    // TODO I think I still need to account for the size of the trailer names.
+    //      Should I be doing that here?
```

Comment:
I think so. I would change the interface on the individual checksum callbacks something like this:
```rust
/// Returns the size in bytes of the actual checksum value
fn checksum_size_hint(&self) -> usize;

/// Returns the checksum header name
fn checksum_header_name(&self) -> HeaderName;

/// Returns the checksum header value
fn checksum_value(&self) -> HeaderValue;
```

The `checksum_value` would only get called once when trailers are needed, but `checksum_header_name` could be called both by trailers and the size hint computation. I think it's safe to make assumptions about the interface here since these are private.